pkg/k8s: do not read k8s node annotations if they are not written

aanm opened the pull request:
When there is an annotation in the k8s node object, the annotation `io.cilium.network.ipv4-cilium-host` is used as the CiliumInternal IP address of the CiliumNode object in [1]. Whenever Cilium is updating any state into the CiliumNode it retrieves all IP address from k8s node, including the ones from annotations, and appends the local node's IP addresses, including the newly correct internal / router IP address, in [2]. Since this is a list, the annotation's IP address is always used first and all other Cilium agents will wrongly use it for any operation.

[1] https://github.com/cilium/cilium/blob/927bd8c26904ff92e42c61cec6d00ea8ac062c05/pkg/nodediscovery/nodediscovery.go#L453-L459
[2] https://github.com/cilium/cilium/blob/927bd8c26904ff92e42c61cec6d00ea8ac062c05/pkg/nodediscovery/nodediscovery.go#L474-L489

Fixes: 73d6cae2c906 ("install: default AnnotateK8sNode to false")
Signed-off-by: André Martins <andre@cilium.io>

```release-note
Fix bug that could lead to inconsistent pod IP information between agents, sometimes leading to a failure to decrypt IPsec traffic.
```

Fixes https://github.com/cilium/cilium/issues/21735

aanm committed:
pkg/k8s: do not read k8s node annotations if they are not written

When there is an annotation in the k8s node object, the annotation
`io.cilium.network.ipv4-cilium-host` is used as the CiliumInternal IP
address of the CiliumNode object in [1]. Whenever Cilium is updating any
state into the CiliumNode it retrieves all IP address from k8s node,
including the ones from annotations, and appends the local node's IP
addresses, including the newly correct internal / router IP
address, in [2]. Since this is a list, the annotation's IP address is
always used first and all other Cilium agents will wrongly use it for
any operation.

[1] https://github.com/cilium/cilium/blob/927bd8c26904ff92e42c61cec6d00ea8ac062c05/pkg/nodediscovery/nodediscovery.go#L453-L459
[2] https://github.com/cilium/cilium/blob/927bd8c26904ff92e42c61cec6d00ea8ac062c05/pkg/nodediscovery/nodediscovery.go#L474-L489

Fixes: 73d6cae2c906 ("install: default AnnotateK8sNode to false")
Signed-off-by: André Martins <andre@cilium.io>

bimmlerd commented on pkg/k8s/node.go:
```diff
@@ -138,6 +111,40 @@ func ParseNode(k8sNode *slim_corev1.Node, source source.Source) *nodeTypes.Node
+	}
+
+	// Any code bellow this line will depend on k8s node annotations. If we are
+	// not annotating the node then we should use any annotation.
```
```suggestion
	// not annotating the node then we should not use any annotations.
```

I think that's what you mean?

bimmlerd reviewed (approved):
LGTM

pchaigno commented on pkg/k8s/node.go:
```diff
@@ -207,7 +214,5 @@ func ParseNode(k8sNode *slim_corev1.Node, source source.Source) *nodeTypes.Node
 	}
 
-	newNode.Labels = k8sNode.GetLabels()
-
```
*Commenting on random line, just to have a resolvable thread.*

Could you also apply the following commit to this branch to expose similar future bugs in the end-to-end tests:
```
commit 2f51308b1dd8d7e7fe8dfca49df9ae4962e7ecf1 (HEAD -> ci-fail-on-router-ip-mismatch)
Author: Paul Chaignon <paul@cilium.io>
Date:   Mon Nov 14 16:11:41 2022 +0100

    test: Fail on router IP mismatch warnings
    
    We try to restore the router IP both from the filesystem (first) and
    from Kubernetes objects (as a fallback). If the two IP addresses don't
    match, we emit a warning.
    
    There is no good reason for this to happen in CI so we should fail the
    test if that warning ever shows up. Doing so would have prevented the
    flake fixed by the previous commit.
    
    Signed-off-by: Paul Chaignon <paul@cilium.io>

diff --git a/test/helpers/cons.go b/test/helpers/cons.go
index 210c11d81d..0d27d9c605 100644
--- a/test/helpers/cons.go
+++ b/test/helpers/cons.go
@@ -229,6 +229,7 @@ const (
        removeTransientRule = "Unable to process chain CILIUM_TRANSIENT_FORWARD with ip" // from https://github.com/cilium/cilium/issues/11276
        missingIptablesWait = "Missing iptables wait arg (-w):"
        localIDRestoreFail  = "Could not restore all CIDR identities" // from https://github.com/cilium/cilium/pull/19556
+       routerIPMismatch    = "Mismatch of router IPs found during restoration"
 
        // ...and their exceptions.
        opCantBeFulfilled        = "Operation cannot be fulfilled on leases.coordination.k8s.io"        // cf. https://github.com/cilium/cilium/issues/16402
@@ -295,6 +296,7 @@ var badLogMessages = map[string][]string{
        removeTransientRule: nil,
        missingIptablesWait: nil,
        localIDRestoreFail:  nil,
+       routerIPMismatch:    nil,
        "DATA RACE":         nil,
        // Exceptions for level=error should only be added as a last resort, if the
        // error cannot be fixed in Cilium or in the test.
```

aanm replied:
Done

pchaigno commented on pkg/k8s/node.go:
```diff
@@ -138,6 +111,40 @@ func ParseNode(k8sNode *slim_corev1.Node, source source.Source) *nodeTypes.Node
+			newNode.EncryptionKey = uint8(u)
+		}
+	}
+
```
*Commenting on random line, just to have a resolvable thread.*

This bug caused a flake because the K8sUpdates test removed the CiliumNode objects as part of `helm delete`. Is there anything we could do so that it also cleans up the Node annotations?

aanm replied:
Filled https://github.com/cilium/cilium/issues/22306

pchaigno reviewed (changes requested):
Discussed offline: this pull request doesn't fix the bug for when `annotate-k8s-node=true`.

aanm committed:
pkg/nodediscovery: do not use Node annotations when mutating CiliumNode

When using CiliumNode, the agent's source of truth should be the agent
itself and not k8s node annotations. Thus we will not use the
annotations for the CiliumInternalIP address when generating a
CiliumNode from the k8s Node resource.

Signed-off-by: André Martins <andre@cilium.io>

pchaigno reviewed (changes requested):
My two comments above are still relevant I think. Other than that, looks good to me.

aanm committed:
test: Fail on router IP mismatch warnings

We try to restore the router IP both from the filesystem (first) and
from Kubernetes objects (as a fallback). If the two IP addresses don't
match, we emit a warning.

There is no good reason for this to happen in CI so we should fail the
test if that warning ever shows up. Doing so would have prevented the
flake fixed by the previous commit.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno reviewed (approved):
I've rerun my local reproduction 8 times with only the second commit and it didn't fail. Given my reproduction works with a rate 1/2, note that there's a probably of 1/256 that it doesn't fail even if it's not fixed.

The changes look good to me.
